libs/vkd3d/resource.c

```diff
@@ -2972,123 +2972,199 @@ static HRESULT d3d12_resource_validate_create_info(const D3D12_RESOURCE_DESC1 *d
     return S_OK;
 }
 
-static HRESULT d3d12_resource_bind_sparse_metadata(struct d3d12_resource *resource,
+static HRESULT d3d12_resource_init_page_table(struct d3d12_resource *resource,
         struct d3d12_device *device, struct d3d12_sparse_info *sparse)
 {
     const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
     VkSparseImageMemoryRequirements *sparse_requirements = NULL;
-    VkSparseImageOpaqueMemoryBindInfo opaque_bind;
+    unsigned int  mip_tail_layer_count, standard_mip_count;
+    VkSparseImageOpaqueMemoryBindInfo opaque_bind_info;
+    unsigned int opaque_bind_count, image_bind_count;
+    VkSparseBufferMemoryBindInfo buffer_bind_info;
     VkTimelineSemaphoreSubmitInfo timeline_info;
+    VkSparseImageMemoryBindInfo image_bind_info;
+    VkImageSubresourceLayers subresource_layers;
+    VkSparseImageMemoryBind *image_binds = NULL;
     VkMemoryRequirements memory_requirements;
-    VkSparseMemoryBind *memory_binds = NULL;
+    VkSparseMemoryBind *opaque_binds = NULL;
     struct vkd3d_queue *vkd3d_queue = NULL;
     uint32_t sparse_requirement_count;
     VkQueue vk_queue = VK_NULL_HANDLE;
-    unsigned int i, j, k, bind_count;
+    VkSparseMemoryBind buffer_bind;
     VkBindSparseInfo bind_info;
     VkDeviceSize metadata_size;
+    unsigned int i, j, k;
     HRESULT hr = S_OK;
     VkResult vr;
 
-    if (d3d12_resource_is_buffer(resource))
-        return S_OK;
-
     /* If a fallback resource, ignore. */
     if (!(resource->flags & VKD3D_RESOURCE_RESERVED))
         return S_OK;
 
-    /* We expect the metadata aspect for image resources to be uncommon on most
-     * drivers, so most of the time we'll just return early. The implementation
-     * is therefore aimed at simplicity, and not very well tested in practice. */
-    VK_CALL(vkGetImageSparseMemoryRequirements(device->vk_device,
-        resource->res.vk_image, &sparse_requirement_count, NULL));
+    memset(&bind_info, 0, sizeof(bind_info));
+    bind_info.sType = VK_STRUCTURE_TYPE_BIND_SPARSE_INFO;
 
-    if (!(sparse_requirements = vkd3d_malloc(sparse_requirement_count * sizeof(*sparse_requirements))))
+    if (d3d12_resource_is_buffer(resource))
     {
-        ERR("Failed to allocate sparse memory requirement array.\n");
-        hr = E_OUTOFMEMORY;
-        goto cleanup;
+        memset(&buffer_bind, 0, sizeof(buffer_bind));
+        buffer_bind.size = resource->desc.Width;
```

Comment:
Do we make sure to round this up to nearest 64 K?

Reply:
```
        /* If we attempt to bind sparse buffer with non-64k pages, we crash drivers.
         * Specs seems a bit unclear how non-aligned VkBuffer sizes are supposed to work,
         * so be safe. Pad out sparse buffers to their natural page size. */
        buffer_info.size = align64(buffer_info.size, VKD3D_TILE_SIZE);
```

Reply:
fixed, wasn't aware that we were padding sparse buffers.